File: OC/instrumentsTest/LeaksTest/LeaksTest/CustomView.m

//
//  CustomView.m
//  LeaksTest
//
//  Created by 江连松 on 2018/11/27.
//  Copyright © 2018年 jls. All rights reserved.
//

#import "CustomView.h"

@interface CustomView()
{
    void(^_block)(NSInteger);
    NSInteger _clickCount;
}
@end

@implementation CustomView

- (instancetype)initWithFrame:(CGRect)frame {
    if (self = [super initWithFrame:frame]) {
        UIButton *button = [UIButton buttonWithType:UIButtonTypeCustom];
        button.frame = CGRectMake(0, 0, 50, 30);
        [button setTitle:@"click" forState:UIControlStateNormal];
        [button addTarget:self action:@selector(clickMe) forControlEvents:UIControlEventTouchUpInside];
        [self addSubview:button];
        
        self.backgroundColor = [UIColor purpleColor];
    }
    return self;
}

- (void)clickMe {
    _clickCount += 1;
    if (_block) {
        _block(_clickCount);
    }
}

- (void)buttonClickCompletion:(void (^)(NSInteger))clickCompletion {
    _block = [clickCompletion copy];
}

@end

File: Swift/DesignPatten/DelegatePattern/DelegatePattern/CustomView.swift
//
//  CustomView.swift
//  DelegatePattern
//
//  Created by JLS on 2017/2/20.
//  Copyright © 2017年 JLS. All rights reserved.
//

import UIKit

protocol HorseDelegate {
    func run(meter: NSInteger)
}

class CustomView: UIView {

    var delegate: HorseDelegate?
    private var meters: NSInteger = 0
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        
        let button = UIButton(type: .custom)
        button.frame = CGRect(x: 100, y: 100, width: 100, height: 100)
        button.center = self.center
        button.setTitle("🐎", for: .normal)
        button.titleLabel?.font = UIFont.systemFont(ofSize: 40)
        button.addTarget(self, action: #selector(buttonAction), for: .touchUpInside)
        self.addSubview(button)
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    func buttonAction() {
        meters += 1
        delegate?.run(meter: meters)
    }
}
